include/deal.II/lac/block_linear_operator.h

```diff
@@ -26,23 +26,365 @@
 
+template <typename Range = BlockVector<double>,
+          typename Domain = Range>
+class BlockLinearOperator;
+
+template <typename Range = BlockVector<double>,
+          typename Domain = Range,
+          typename BlockMatrix>
+BlockLinearOperator<Range, Domain>
+block_operator(const BlockMatrix &matrix);
+
+template <size_t m, size_t n,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<Range, Domain>
+block_operator(const std::array<std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, n>, m> &);
+
+template <size_t m,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<Range, Domain>
+block_diagonal_operator(const std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, m> &);
+
+template <size_t m,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<Range, Domain>
+block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename Domain::BlockType> &op);
+
+// This is a workaround for a bug in <=gcc-4.7 that does not like partial
+// template default values in combination with local lambda expressions [1]
+//
+// [1] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53624
+//
+// Forward declare functions with partial template defaults:
+
+template <typename Range = BlockVector<double>,
+          typename Domain = Range,
+          typename BlockMatrix>
+BlockLinearOperator<Range, Domain>
+block_diagonal_operator(const BlockMatrix &block_matrix);
+
+template <typename Range = BlockVector<double>,
+          typename Domain = Range>
+LinearOperator<Domain, Range>
+block_forward_substitution(const BlockLinearOperator<Range, Domain> &,
+                           const BlockLinearOperator<Domain, Range> &);
+
+template <typename Range = BlockVector<double>,
+          typename Domain = Range>
+LinearOperator<Domain, Range>
+block_back_substitution(const BlockLinearOperator<Range, Domain> &,
+                        const BlockLinearOperator<Domain, Range> &);
+
+// end of workaround
+
 /**
- * @name Creation of LinearOperator block structures
+ *  TODO: Description
+ */
+template <typename Range, typename Domain>
+class BlockLinearOperator : public LinearOperator<Range, Domain>
+{
+public:
+
+  typedef LinearOperator<typename Range::BlockType, typename Domain::BlockType> BlockType;
+
+  /**
+   * Create an empty BlockLinearOperator object.
+   *
+   * All<code>std::function</code> member objects of this class and its
+   * base class LinearOperator are initialized with default variants that
+   * throw an exception upon invocation.
+   */
+  BlockLinearOperator()
+    : LinearOperator<Range, Domain>()
+  {
+
+    n_block_rows = []() -> unsigned int
+    {
+      Assert(false, ExcMessage("Uninitialized BlockLinearOperator<Range, "
+      "Domain>::n_block_rows called"));
```

Comment:
Not important from my perspective. As long as it's consistent I'm ok with it. Some future astyle version will learn to deal with this.